### PR TITLE
Add symbol's assumptions to the global_assumptions context

### PR DESCRIPTION
**Pull request by skirpichev**
``` python
In [2]: r = Symbol('r', real=True)
In [3]: x = Symbol('x')
In [4]: ask(Q.real(x))
In [5]: ask(Q.real(r))
Out[5]: True
In [6]: ask(Q.real(Symbol('spam')))
In [7]: ask(Q.real(Symbol('spam', real=True)))
Out[7]: True
In [8]: ask(Q.real(Symbol('spam')))
```

### CODE REVIEW

**Comment by skirpichev** [comment]
Ok, tests pass now.  This is just a regular PR, please review.

About weakref, this should be doable.  But right now this doesn't work for me with cache.  I'll post an issue.

**Comment by skirpichev** [comment]
cache bug (I think) - #8825

**Comment by asmeurer** [comment]
How do you feel about this vs. #7925?

**Comment by skirpichev** [comment]
This solution looks to be more simple.

Also, I don't think it's a good idea to add a new storage for facts.  We have assumptions from the context manager, and we have local assumptions from the second ask argument.

**Comment by skirpichev** [comment]
rebased

**Comment by asmeurer** [comment]
> Also, I don't think it's a good idea to add a new storage for facts. We have assumptions from the context manager, and we have local assumptions from the second ask argument.

I don't understand what you are referring to here.

**Comment by skirpichev** [comment]
> I don't understand what you are referring to here.

You add additional (third) storage for assumptions (every Symbol contains its own assumptions).

**Comment by skirpichev** [comment]
./bin/test sympy/core runs in 307.40s for this branch (vs 294.05 in the master).

**Comment by asmeurer** [comment]
I still don't follow. My pull request doesn't touch the core. Symbol already contains its own (old) assumptions.

**Comment by skirpichev** [comment]
On Thu, Jan 15, 2015 at 09:29:57AM -0800, Aaron Meurer wrote:
>    Symbol already contains its own (old) assumptions.

Yes.  But this storage was "hidden" from the new assumptions.

**Comment by asmeurer** [comment]
OK, I understand what you are saying now. 

Something to note is that your pull request already implicitly relies on this behavior, because `Symbol('x') != Symbol('x', real=True)`. If that weren't the case, then this would be very surprising behavior. Creating a Symbol with assumptions would change the global assumptions for every symbol of that name. 

As I've stated before, I think the primary difference between your pull request and mine is that yours is eager (every Symbol that is created with assumptions is added to the new assumptions system when it is created), whereas mine is lazy (the new assumptions only look at the Symbol (old) assumptions when they are used). My approach is likely more efficient when the assumptions are rarely used (which they currently are). Yours is likely more efficient if they are used heavily, as using a fact is faster than using a handler. Hence, I, at least for now, prefer my approach, because it's new code that is only ever called when the new assumptions are actually used.

**Comment by skirpichev** [comment]
>    Something to note is that your pull request already implicitly relies on
>    this behavior, because Symbol('x') != Symbol('x', real=True). If that
>    weren't the case, then this would be very surprising behavior.

Sure, it's.

>    Creating a
>    Symbol with assumptions would change the global assumptions for every
>    symbol of that name.

It's not what we currently have in this PR, but yes.
When we remove old assumptions and deprecate old storage (assumptions0
property and so on) - Symbols and assumptions will be separate.  Do you
think it's bad idea?

Right now you have funny issues with printing for Symbol('x') vs
Symbol('x', real=True).  But it that a good idea at all to have
assumptions, attached to Symbol's?

My approach (huh, I think it's actually Mathematica's one ;) seems
to be more flexible.  You could attach/detach assumptions.  Symbols
become just name holders.

>    As I've stated before, I think the primary difference between your pull
>    request and mine is that yours is eager (every Symbol that is created with
>    assumptions is added to the new assumptions system when it is created),
>    whereas mine is lazy (the new assumptions only look at the Symbol (old)
>    assumptions when they are used).

The main difference, as I said, is that you introduce a new
storage for new assumptions system (right now we have AssumptionsContext
and assumptions kwarg for ask).

>    My approach is likely more efficient when
>    the assumptions are rarely used (which they currently are).

Any proofs?

>    Yours is
>    likely more efficient if they are used heavily, as using a fact is faster
>    than using a handler.

I don't see the difference.  Can you suggest better test?

>    Hence, I, at least for now, prefer my approach,
>    because it's new code that is only ever called when the new assumptions
>    are actually used.

Shouldn't you argue first your efficiency points and so on?

**Comment by asmeurer** [comment]
> But it that a good idea at all to have
assumptions, attached to Symbol's?

No not all the assumptions, but I do think there is value in allowing it.

> Any proofs?

My code is not run when Symbols are created and yours is. 

> I don't see the difference.  Can you suggest better test?

I didn't benchmark this one, so maybe it isn't noticeable, but the algorithm checks if it can solve things from the facts alone first, and then falls back to the handlers if it can't. 

I wouldn't want to do any premature optimization with this, especially since I don't really like the handler system anyway, but I'd say in general, the assumptions system should check things in tiers, starting with things that are easy and fast, but maybe not so powerful and moving on to more complicated and potentially slower algorithms. And in general, things should go in the earliest tier that they can go in. So for instance, if a fact can be written entirely using predicates without any reference to the objects themselves, it should go in the known_facts dictionary. 

But this is mostly thinking ahead to a more ideal architecture for the new assumptions. The way that we implement this doesn't really matter too much, because it will be easy to adapt to whatever way we end up doing things in the future. 

I just thought of another potential advantage of your way (although I didn't test it yet). If I do `assume(Q.positive(Symbol('x', negative=True))` with your way it should detect the inconsistency, but with my way it won't (possibly ever, because the handlers will not be called if it can deduce things from the global assumptions). 

So to summarize, I actually don't like handlers, so my primary concern with this pull request is the performance and memory implications of doing this with every Symbol creation (and especially [this](https://github.com/sympy/sympy/pull/7926#issuecomment-53586927)).

**Comment by skirpichev** [comment]
> I didn't benchmark this one, so maybe it isn't noticeable

I did.  But maybe core tests - not the best variant.  For now it is what it is - I don't see measurable difference.

> and memory implications of doing this with every Symbol creation (and especially this)

In principle, this shouldn't be too hard to implement AssumptionsContext with weakref's, but we have here a cache issue, see #8825.

**Comment by skirpichev** [comment]
Feel free to pick up this work.
